Introducing Socialised Stories Updates

kumarpalsinh25 opened the pull request:
Fixes, https://github.com/acterglobal/a3/issues/2502

Note to reviewers : 
- This PR gets slightly bigger but there were no other option to reduce down this work in smaller pieces so let me try to explain as much as I can this PR Description and Code comments.
- Many of changes are related to Naming changes.

### Introducing new type of Acter Update which is called **Stories**

![Acter Updates](https://github.com/user-attachments/assets/a56e7866-d049-496e-aec8-4d52dae2be92)

- - -
- Technically Boost and Story are mostly same expect below key difference.
**STORIES :** "Everyone can see and It will disappears in 14 days",
**BOOST :** "Important News. Which also sends a push notification to space members",

<img width="564" alt="Screenshot 2025-02-21 at 11 06 34 AM" src="https://github.com/user-attachments/assets/6b41bd8f-8351-4612-977f-68b267b72bfa" />

- - -
### Summary of changes

### 1) Removed space selection from post editor screen:
As Boost/Stories action items are independent from space selection, I have remove space selection option from editor window.

https://github.com/user-attachments/assets/bc91579e-b4d8-4234-8559-89579e7a6791

### 2) Introducing all new `PostTo` screen:
As now you there are two type of Updates which is Stories and Boost, so upon setting update different slides from post editor you will be navigating to new **PostTo** screen where you need to do,
- Space selection
- Post Type selection which is Stories or Boost (This options are enabled based on the selected space permissions)


https://github.com/user-attachments/assets/8e6be541-374b-408f-b5eb-96526d39a1e2




### 3) Changes in View Updates:
- Updates filter tab added at the top to filters it based on the type.
- For Boost Update : you will will space avatar in bottom right and Space Name and Post time in bottom left side
- For Story Update : you will will user avatar (who posted story) in bottom right and User Name and Space name in bottom left side.


https://github.com/user-attachments/assets/362333a8-2f39-4cfa-9ae2-0478a15ea2f6

https://github.com/user-attachments/assets/49cb2a79-a38a-4d25-8dcc-eaf26f46f602



### 4) Overview of key code changes:
- Created different abstract classes to combine news and story related stuff in one.

kumarpalsinh25 commented on app/integration_test/tests/updates.dart:
```diff
@@ -33,16 +33,15 @@ extension ActerNews on ConvenientTest {
   Future<void> toggleBackgroundColor() async {
-    final slideBackgroundColorKey =
-        find.byKey(NewsUpdateKeys.slideBackgroundColor);
+    final slideBackgroundColorKey = find.byKey(UpdateKeys.slideBackgroundColor);
```
Only name changes from `NewsUpdateKeys` to `UpdateKeys`

kumarpalsinh25 commented on app/integration_test/tests/updates.dart:
```diff
@@ -72,39 +71,39 @@ extension ActerNews on ConvenientTest {
     );
-    final slide = NewsSlideItem(
-      type: NewsSlideType.image,
+    final slide = UpdateSlideItem(
```
Only name changes from `NewsSlideItem` to `UpdateSlideItem`

kumarpalsinh25 commented on app/lib/common/utils/routes.dart:
```diff
@@ -28,6 +28,7 @@ enum Routes {
   update('/updates/:updateId([^/]+)'),
   updateList('/updateList'),
   actionAddUpdate('/actions/addUpdate'),
+  actionAddUpdatePostTo('/actions/addUpdate/PostTo'),
```
Added new `PostTo screen `as show-case in PR Description

kumarpalsinh25 replied:
Understood, let me remove this route and make it based on the bottom sheet within same editor.

kumarpalsinh25 replied:
Done with https://github.com/acterglobal/a3/pull/2537/commits/5ac655149f2ec3ff06c51d9136ed2e008f2f2b24

Reference Video:

https://github.com/user-attachments/assets/12fe15bb-8ee1-4b4e-badb-4b6220e74941

kumarpalsinh25 commented on app/lib/features/comments/types.dart:
```diff
@@ -57,11 +58,11 @@ extension TaskCommentsManagerProviderExtension on Task {
 
-class NewsEntryCommentsManagerProvider implements CommentsManagerProvider {
-  final NewsEntry inner;
+class UpdateEntryCommentsManagerProvider implements CommentsManagerProvider {
```
Naming changes only.

kumarpalsinh25 commented on app/lib/features/news/actions/make_image_slide.dart:
```diff
@@ -0,0 +1,78 @@
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:mime/mime.dart';
+
+Future<NewsSlideDraft> makeImageSlideForNews(
```
No logical changes. Just copied logic from previous implementation and used it based on the post type.

kumarpalsinh25 commented on app/lib/features/news/actions/make_text_slide.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+Future<NewsSlideDraft> makeTextSlideForNews(
```
No logical changes. Just copied logic from previous implementation and used it based on the post type.

kumarpalsinh25 commented on app/lib/features/news/actions/make_video_slide.dart:
```diff
@@ -0,0 +1,70 @@
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:mime/mime.dart';
+
+Future<NewsSlideDraft> makeVideoSlideForNews(
```
No logical changes. Just copied logic from previous implementation and used it based on the post type.

kumarpalsinh25 commented on app/lib/features/news/actions/submit_story.dart:
```diff
@@ -0,0 +1,89 @@
+
+final _log = Logger('a3::updates::submit_story');
+
+Future<void> sendStory(BuildContext context, WidgetRef ref) async {
```
No logical changes. Just copied logic from previous implementation and used it for story post type.

kumarpalsinh25 commented on app/lib/features/news/model/news_post_state.freezed.dart:
```diff
@@ -16,8 +16,8 @@ final _privateConstructorUsedError = UnsupportedError(
 mixin _$NewsPostState {
-  NewsSlideItem? get currentNewsSlide => throw _privateConstructorUsedError;
-  List<NewsSlideItem> get newsSlideList => throw _privateConstructorUsedError;
+  UpdateSlideItem? get currentUpdateSlide => throw _privateConstructorUsedError;
```
This entire class will only contains naming changes, no change in logic.

kumarpalsinh25 commented on app/lib/features/news/model/type/update_entry.dart:
```diff
@@ -0,0 +1,130 @@
+
+bool isStory(UpdateEntry inner) => inner is UpdateStoryEntry;
+
+abstract class UpdateEntry {
```
Generalised abstract class creation.

kumarpalsinh25 commented on app/lib/features/news/model/type/update_slide.dart:
```diff
@@ -0,0 +1,65 @@
+import 'package:acter_flutter_sdk/acter_flutter_sdk_ffi.dart';
+
+abstract class UpdateSlide {
```
Generalised abstract class creation.

kumarpalsinh25 commented on app/lib/features/news/pages/add_news/add_news_post_to_page.dart:
```diff
@@ -0,0 +1,199 @@
+
+enum PostTypeSelection { story, boost, none }
+
+class AddNewsPostToPage extends ConsumerStatefulWidget {
```
All new PostTo page for space selection and post type selection.

kumarpalsinh25 commented on app/lib/features/news/providers/news_post_editor_providers.dart:
```diff
@@ -31,15 +31,15 @@ class NewsStateNotifier extends StateNotifier<NewsPostState> {
-    NewsSlideItem? selectedNewsSlide = state.currentNewsSlide;
-    selectedNewsSlide?.backgroundColor = getRandomElement(newsPostColors);
-    state = state.copyWith(currentNewsSlide: selectedNewsSlide);
+    UpdateSlideItem? selectedUpdateSlide = state.currentUpdateSlide;
```
Naming changes only.

kumarpalsinh25 commented on app/lib/features/news/providers/news_providers.dart:
```diff
@@ -1,28 +1,88 @@
 import 'package:riverpod/riverpod.dart';
 
+//EVENT FILTERS
+enum UpdateFilters {
```
Added filters for updates

kumarpalsinh25 commented on app/lib/features/news/widgets/news_item_slide/news_filter_buttons.dart:
```diff
@@ -0,0 +1,51 @@
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class NewsFilterButtons extends ConsumerWidget {
```
Updates Filter Buttons UI.

<img width="237" alt="Screenshot 2025-02-21 at 11 51 30 AM" src="https://github.com/user-attachments/assets/3bc608b6-002a-4470-bc2f-5bbf85ea4a72" />

gnunicorn replied:
can we have this a little more subtle? It is taking up a _lot_ of space and attention in the updates screen right now: just do some picture posts rather than just one-line-texts and you'll notice it is pretty intense and in the way.


I was thinking more of the way that tiktok does it: one-liner, simple (no padding or borders), selected is bold and/or underlined:

![image](https://github.com/user-attachments/assets/b153dc42-89a5-48c8-be73-e39ba9e78d51)

kumarpalsinh25 replied:
Yes sure, make sense! will change it accordingly.

kumarpalsinh25 replied:
@gnunicorn 
Here is the updated design based on your thoughts.


https://github.com/user-attachments/assets/29ce2a1a-7a69-4952-b95a-d9c163b4090a

gnunicorn commented on native/acter/src/api/room.rs:
```diff
@@ -226,8 +226,9 @@ impl Member {
-                    return false;
+                    // Story is less restricted than news
+                    // Everyone can post story as default
+                    return true;
```
wait. no. this still needs to be activated and if there are no permissions set, it isn't to be understood as activated. this is false by  default, too.
```suggestion
                    return false;
```

kumarpalsinh25 replied:
Oh then I need to add code in Space Setting >> Space Feature >> Stories, correct?

<img width="400" alt="Screenshot 2025-02-26 at 1 08 32 PM" src="https://github.com/user-attachments/assets/8b025649-a820-4815-8da9-0a9a8b5eca5c" />

<img width="389" alt="Screenshot 2025-02-26 at 1 08 54 PM" src="https://github.com/user-attachments/assets/92e4ba01-8145-4e4b-a4fd-e28943922690" />

Also need to add action button for Stories in Space Overview page?

<img width="418" alt="Screenshot 2025-02-26 at 1 09 24 PM" src="https://github.com/user-attachments/assets/2a13d735-1b44-4da9-b3f2-6927b9dde673" />

kumarpalsinh25 replied:
@gnunicorn 
Can't we make stories as auto-activated as many people still doesn't know about this things?

gnunicorn replied:
> Oh then I need to add code in Space Setting >> Space Feature >> Stories, correct?

Yes, we need to add the feature in the space settings as well. Same for allowing to create them from a space, yes.

> Can't we make stories as auto-activated as many people still doesn't know about this things?


No. We don't overrule the features people have activated when rolling out new features. The argument "but people don't know about it" would always apply to all new features but as we don't give admins any way to _disable_ this behavior beforehand, what gives us the right to impose the feature they should have activated in their existing spaces.

Now for spaces created _after_ we introduced stories, this will be different: once rolled out and considered working fine, we will make this (as well as other features) default on upon creation as part of the creation process, but it is something the creators _actively_ agree to. See https://balsamiq.cloud/sj2u1jx/plw5law .

kumarpalsinh25 replied:
Got it, let me add required code in Space Settings as well as Space Overview screen.

kumarpalsinh25 replied:
Done with below commits.

### 1) Add Story action button in Space Overview screen:
https://github.com/acterglobal/a3/pull/2537/commits/f9a4bd4fc19c41de925a87913afb31c5442e0541
![Screenshot 2025-02-27 at 3 00 48 PM](https://github.com/user-attachments/assets/32203f39-5afc-4b40-a078-17e3cc76c678)


### 2) Added stories activation option in Space Feature settings:
https://github.com/acterglobal/a3/pull/2537/commits/75b43d9f69a238c60914a910015058943d30177a
![Screenshot 2025-02-27 at 3 01 43 PM](https://github.com/user-attachments/assets/56655f04-9605-4a57-ad96-b35e45c7d73f)
![Screenshot 2025-02-27 at 3 01 55 PM](https://github.com/user-attachments/assets/87bc002f-509e-4ba6-b063-4dffd1e48964)

kumarpalsinh25 commented on app/lib/common/widgets/room/brief_room_list_entry.dart:
```diff
@@ -8,6 +8,7 @@ class BriefRoomEntry extends ConsumerWidget {
   final String roomId;
   final String? selectedValue;
   final String canCheck;
+  final String? secondaryCanCheck;
   final String keyPrefix;
   final Function(String)? onSelect;
```
Sorry but not getting you here. `Function(Membership)? canCheck` callback function to verify 2 permission?

gnunicorn replied:
yes. It's would be a callback that checked for all or any permissions, letting the outer widget decide on the specifics rather than encoding it in here (which is encoded as either-or right now)

gnunicorn replied:
fixed with https://github.com/acterglobal/a3/pull/2537/commits/152661c8abfd325e37ff205ca13b5f0c599408ab

kumarpalsinh25 commented on app/lib/common/widgets/room/brief_room_list_entry.dart:
```diff
@@ -27,27 +29,37 @@ class BriefRoomEntry extends ConsumerWidget {
       orElse: () => false,
     );
+
+    if (secondaryCanCheck != null) {
+      bool seocndayrCanPermission = roomMembership.maybeWhen(
+        data: (membership) => membership?.canString(secondaryCanCheck!) == true,
+        orElse: () => false,
```
Sorry, which typo mistakes ?

gnunicorn replied:
`seocndayrCanPermission` -> `secondaryCanPermission`

two typos in the same variable.

kumarpalsinh25 commented on app/lib/common/widgets/space_name_widget.dart:
```diff
@@ -5,11 +5,13 @@ import 'package:flutter_riverpod/flutter_riverpod.dart';
 class SpaceNameWidget extends ConsumerWidget {
   final String spaceId;
   final bool isShowBrackets;
+  final bool isShowUnderline;
```
I just followed the naming of `isShowBrackets` and make start it with `is` to highlight it as bool param but let me change it.

kumarpalsinh25 replied:
Done with https://github.com/acterglobal/a3/pull/2537/commits/cc71d179da81f904da80a260de254ed9bcfdb0a4

kumarpalsinh25 commented on app/lib/features/news/pages/add_news/add_news_post_to_page.dart:
```diff
@@ -0,0 +1,203 @@
+import 'package:acter/common/providers/room_providers.dart';
```
Agree, let me write tests for this page.

kumarpalsinh25 replied:
Done with https://github.com/acterglobal/a3/pull/2537/commits/de0ff9462c4e62348c8ac380a1c48298df3fc97d

kumarpalsinh25 commented on app/lib/features/news/pages/add_news/add_news_post_to_page.dart:
```diff
@@ -0,0 +1,203 @@
+enum PostTypeSelection { story, boost, none }
+
+class AddNewsPostToPage extends ConsumerStatefulWidget {
+  final String? initialSelectedSpace;
```
Oh ya, let me remove `initialSelectedSpace` as things are managed by newsState providers here.

kumarpalsinh25 commented on app/lib/features/news/pages/add_news/add_news_post_to_page.dart:
```diff
@@ -0,0 +1,203 @@
+        Padding(
+          padding: const EdgeInsets.all(18),
+          child: ActerPrimaryActionButton(
+            onPressed: () {
+              if (selectedSpaceId == null || selectedSpaceId.isEmpty) {
+                EasyLoading.showToast(lang.pleaseSelectSpace);
+              } else if (canPostBoost.value == false &&
+                  canPostStories.value == false) {
+                EasyLoading.showToast(lang.notHaveBoostStoryPermission);
+              } else if (selectedPostType.value == PostTypeSelection.none) {
+                EasyLoading.showToast(lang.pleaseSelectePostType);
+              } else if (selectedPostType.value == PostTypeSelection.story) {
+                sendStory(context, ref);
+              } else if (selectedPostType.value == PostTypeSelection.boost) {
+                sendNews(context, ref);
+              }
```
Yes sure, let me move it to local function.

kumarpalsinh25 replied:
Done with https://github.com/acterglobal/a3/pull/2537/commits/7b19e80b6a8fa38808996f2cbb1f976fdd0f3d8a

kumarpalsinh25 commented on app/lib/features/news/pages/add_news/add_news_post_to_page.dart:
```diff
@@ -0,0 +1,203 @@
+  Widget spaceSelector() {
+    final newsPostSpaceId = ref.watch(newsStateProvider).newsPostSpaceId;
+
+    final spaceSelectorWidget = (newsPostSpaceId != null)
```
There were things due to which I created it separate and not used common one.

1) Need to make change in common space selector `canCheck` and I need to check check 2 permissions `canPostNews` and `canPostStories` for space selection.
2) I need to update `spaceId` in  final notifier = ref.read(newsStateProvider.notifier) so that I need to either update `SelectSpaceFormField` with `spaceId` callback or need to use `ref.read(selectedSpaceIdProvider)` to update spaceId in `newsStateProvider`.
3) And last thing, I was planing slightly different design when space is actually selected.

Happy to use common `SelectSpaceFormField` widget by making above changes over there.

Note: I have already used common `selectSpaceDrawer`, but just haven't used common `SelectSpaceFormField` due to above reasons.

kumarpalsinh25 replied:
Make sense, will do changes accordingly.

kumarpalsinh25 replied:
I have tried to use common `SelectSpaceFormField` here after making necessary changes but there are multiple issues I am facing when I use common space selector, is this really necessary?

because adding different bool flag to manage navigation other other stuff will make `SelectSpaceFormField` bit complexed.


https://github.com/user-attachments/assets/97bbefaf-35e9-4d61-9ed4-a854e8832c0e

kumarpalsinh25 replied:
Done with..

- https://github.com/acterglobal/a3/pull/2537/commits/33371819d270ec531f9304a1fb1340eb1428cd6d#diff-c4e384386deb24ddabf12d09c85795a9ec1e97af0b7e34959cea734a5537deab
- https://github.com/acterglobal/a3/pull/2537/commits/8bebad06612d78dbde9ca561165e3e0ff6f2a780

Reference Video:

https://github.com/user-attachments/assets/bdcfdcd0-2150-4dc2-b4eb-769e3f437f44

kumarpalsinh25 commented on app/lib/features/news/providers/notifiers/story_list_notifier.dart:
```diff
@@ -0,0 +1,62 @@
+    _poller = _listener.listen(
+      (data) async {
+        _log.info('stories subscribe received');
+        state = await AsyncValue.guard(() => _fetchStories(client, spaceId));
```
Code it just copy pasted from `AsyncNewsListNotifier`, but surely let me update code accordingly in both places.

kumarpalsinh25 commented on app/lib/features/news/widgets/news_item/news_item.dart:
```diff
@@ -46,31 +50,71 @@ class _NewsItemState extends ConsumerState<NewsItem> {
+              child: Row(
+                children: [
+                  Text(L10n.of(context).inSpaceLabelInline),
+                  SizedBox(width: 8),
```
Yes sure

kumarpalsinh25 replied:
Done with https://github.com/acterglobal/a3/pull/2537/commits/5ba13df12b5e609c1ed6fb8d183a37215ed32fa6

Reference Image:

![Screenshot 2025-02-26 at 4 04 34 PM](https://github.com/user-attachments/assets/3287226e-39c7-45c6-8042-66b20db5b228)

kumarpalsinh25 commented on app/lib/l10n/app_en.arb:
```diff
@@ -2354,6 +2354,13 @@
   "unableToLoadVideo": "Unable to load video",
   "unableToLoadImage": "Unable to load image",
+  "story": "Story",
+  "storyInfo": "Everyone can see, this is from you. It disappears in 14 days",
```
Let me update text accordingly.

kumarpalsinh25 replied:
Done with https://github.com/acterglobal/a3/pull/2537/commits/28fc21e1b749b2a65f23d685507ef50ac3f2c3d5

kumarpalsinh25 commented:
@gnunicorn 
I have addresses all the PR Feedback points and marked every comment as resolved by adding relevant git commit and Screenshots/Video Reference details.

gnunicorn commented:
@kumarpalsinh25 great work. just some CI is still failing.
